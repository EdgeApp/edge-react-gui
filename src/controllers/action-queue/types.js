// @flow
/* eslint-disable no-use-before-define */

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'exchange-buy'
  | 'exchange-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type ActionOpExecStatus = 'pending' | 'active' | 'done' | Error

export type ActionOp =
  | {
      type: 'seq',
      actions: ActionOp[]
    }
  | {
      type: 'par',
      actions: ActionOp[]
    }
  | {
      type: 'broadcast-tx',
      pluginId: string,
      rawTx: Uint8Array
    }
  | {
      type: 'exchange-buy',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'exchange-sell',
      nativeAmount: string,
      walletId: string,
      tokenId?: string,
      exchangePluginId: string
    }
  | {
      type: 'loan-borrow',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-deposit',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-repay',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'loan-withdraw',
      borrowPluginId: string,
      nativeAmount: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'swap',
      fromWalletId: string,
      toWalletId: string,
      fromTokenId?: string,
      toTokenId?: string,
      nativeAmount: string,
      amountFor: 'from' | 'to'
    }
  // Useful for development/testing
  | {
      type: 'toast',
      message: string
    }
  | {
      type: 'delay',
      ms: number
    }

//
// Action (After) Effects
//

export type ActionEffect =
  | {
      type: 'seq',
      opIndex: number,
      childEffect: ActionEffect
    }
  | {
      type: 'par',
      childEffects: ActionEffect[]
    }
  | {
      type: 'address-balance',
      address: string,
      aboveAmount?: string,
      belowAmount?: string,
      walletId: string,
      tokenId?: string
    }
  | {
      type: 'tx-confs',
      txId: string,
      walletId: string,
      confirmations: number
    }
  | {
      type: 'price-level',
      currencyPair: string,
      aboveRate?: number,
      belowRate?: number
    }
  | {
      type: 'done',
      error?: Error
    }
  // Useful for development/testing
  | {
      type: 'unixtime',
      timestamp: number
    }
  | {
      type: 'init'
    }

//
// Action Program
//

// On Disk:
export type ActionProgram = {
  programId: string,
  actionOp: ActionOp
}

export type ActionProgramState = {
  programId: string,
  effect?: ActionEffect
}

//
// Internal Types
//

export type ActionQueueItem = {
  program: ActionProgram,
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

export type ExecutionResult = {
  effect: ActionEffect,
  action?: ActionOp // For dryrun
}
export type ExecutionResults = {
  nextState: ActionProgramState
}

//
// Aciton Display API
//

export type ActionDisplayInfo = {
  title: string,
  message: string,
  status: ActionOpExecStatus,
  steps: ActionDisplayInfo[]
}
