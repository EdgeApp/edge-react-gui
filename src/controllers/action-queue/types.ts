/* eslint-disable no-use-before-define */

import { EdgeAccount, EdgeNetworkFee, EdgeTransaction } from 'edge-core-js'

//
// Action Operations
//
export type ActionOpTypes =
  | 'seq'
  | 'par'
  | 'wyre-buy'
  | 'wyre-sell'
  | 'loan-borrow'
  | 'loan-deposit'
  | 'loan-repay'
  | 'loan-withdraw'
  | 'swap'
  | 'toast'
  | 'delay'

export type SeqActionOp = {
  type: 'seq'
  actions: ActionOp[]
}
export type ParActionOp = {
  type: 'par'
  actions: ActionOp[]
}
export type BroadcastTxActionOp = {
  type: 'broadcast-tx'
  pluginId: string
  rawTx: Uint8Array
}
export type WyreBuyActionOp = {
  type: 'wyre-buy'
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type WyreSellActionOp = {
  type: 'wyre-sell'
  wyreAccountId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanBorrowActionOp = {
  type: 'loan-borrow'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanDepositActionOp = {
  type: 'loan-deposit'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanRepayActionOp = {
  type: 'loan-repay'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type LoanWithdrawActionOp = {
  type: 'loan-withdraw'
  borrowPluginId: string
  nativeAmount: string
  walletId: string
  tokenId?: string
}
export type SwapActionOp = {
  type: 'swap'
  amountFor: 'from' | 'to'
  fromTokenId?: string
  fromWalletId: string
  nativeAmount: string
  toTokenId?: string
  toWalletId: string
}
export type ActionOp =
  | SeqActionOp
  | ParActionOp
  | BroadcastTxActionOp
  | WyreBuyActionOp
  | WyreSellActionOp
  | LoanBorrowActionOp
  | LoanDepositActionOp
  | LoanRepayActionOp
  | LoanWithdrawActionOp
  | SwapActionOp

//
// Action (After) Effects
//

export type SeqEffect = {
  type: 'seq'
  opIndex: number
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type ParEffect = {
  type: 'par'
  childEffects: Array<ActionEffect | null> // null is only for dryrun
}
export type AddressBalanceEffect = {
  type: 'address-balance'
  address: string
  aboveAmount?: string
  belowAmount?: string
  walletId: string
  tokenId?: string
}
export type PushEventEffect = {
  type: 'push-event'
  eventId: string
}
export type PriceLevelEffect = {
  type: 'price-level'
  currencyPair: string
  aboveRate?: number
  belowRate?: number
}
export type TxConfsEffect = {
  type: 'tx-confs'
  txId: string
  walletId: string
  confirmations: number
}
export type DoneEffect = {
  type: 'done'
  error?: Error
  cancelled?: boolean
}

export type ActionEffect = SeqEffect | ParEffect | AddressBalanceEffect | PushEventEffect | PriceLevelEffect | TxConfsEffect | DoneEffect

//
// Action Program
//

// Storage:
export type ActionProgram = {
  programId: string
  actionOp: ActionOp
  // Development mode flag
  mockMode?: boolean
}
export type ActionProgramState = {
  clientId: string
  programId: string
  effect?: ActionEffect

  // Flags:
  effective: boolean // Whether the effect is observed
  executing: boolean // Whether the program is executing

  lastExecutionTime: number // The time when the effect was checked
  nextExecutionTime: number // The next time when the effect should be checked again
}

export type ActionQueueItem = {
  program: ActionProgram
  state: ActionProgramState
}
export type ActionQueueMap = {
  [id: string]: ActionQueueItem
}

// Runtime:
export type BroadcastTx = {
  walletId: string
  networkFee: EdgeNetworkFee
  tx: EdgeTransaction
}
export type ExecutableAction = {
  dryrunOutput: ExecutionOutput | null
  execute: () => Promise<ExecutionOutput>
}
export type ExecutionContext = {
  account: EdgeAccount
  clientId: string
}
export type ExecutionOutput = {
  effect: ActionEffect
  broadcastTxs: BroadcastTx[]
}
export type ExecutionResults = {
  nextState: ActionProgramState
}
export type PendingTxMap = {
  [walletId: string]: EdgeTransaction[]
}

//
// Aciton Display API
//

export type ActionDisplayStatus = 'pending' | 'active' | 'done' | Error

export type ActionDisplayInfo = {
  title: string
  message: string
  status: ActionDisplayStatus
  steps: ActionDisplayInfo[]
}
