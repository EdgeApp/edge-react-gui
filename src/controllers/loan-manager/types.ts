import { BorrowEngine, BorrowPlugin } from '../../plugins/borrow-plugins/types'
import { LoanProgramEdge } from './store'

export type LoanAccount = {
  id: string
  borrowPlugin: BorrowPlugin
  borrowEngine: BorrowEngine

  // Action Program Relationships
  programEdges: LoanProgramEdge[]
}

export type LoanAccountMap = {
  [walletId: string]: LoanAccount
}

export type LoanManager = {
  loanAccounts: LoanAccount[]
}
