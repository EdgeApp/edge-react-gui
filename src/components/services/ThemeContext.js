// @flow

import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { womanTheme } from '../../theme/variables/womanTheme'
import { type Theme } from '../../types/Theme.js'

export type { Theme }
export { cacheStyles }

/**
 * Utility type for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(womanTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
