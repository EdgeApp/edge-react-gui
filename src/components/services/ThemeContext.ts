import { cacheStyles, makeThemeContext } from 'react-native-patina'

import { config } from '../../theme/appConfig'
import { Theme } from '../../types/Theme'

export type { Theme }
export { cacheStyles }

/**
 * Utility for declaring `withTheme` components.
 */
export type ThemeProps = {
  theme: Theme
}

// Provide the theme context methods:
const themeContext = makeThemeContext(config.darkTheme)
export const { ThemeProvider, useTheme, withTheme, changeTheme, getTheme, watchTheme } = themeContext
