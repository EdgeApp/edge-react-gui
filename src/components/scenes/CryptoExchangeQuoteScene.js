// @flow

import { div, gte } from 'biggystring'
import { type EdgeAccount } from 'edge-core-js/types'
import * as React from 'react'
import { ScrollView, View } from 'react-native'
import FastImage from 'react-native-fast-image'

import { exchangeTimerExpired, shiftCryptoCurrency } from '../../actions/CryptoExchangeActions'
import s from '../../locales/strings.js'
import { Slider } from '../../modules/UI/components/Slider/Slider'
import { connect } from '../../types/reactRedux.js'
import { type RouteProp } from '../../types/routerTypes.js'
import { type GuiSwapInfo } from '../../types/types.js'
import { getSwapPluginIconUri } from '../../util/CdnUris'
import { logEvent } from '../../util/tracking.js'
import { CircleTimer } from '../common/CircleTimer'
import { SceneWrapper } from '../common/SceneWrapper.js'
import { ButtonsModal } from '../modals/ButtonsModal'
import { swapVerifyTerms } from '../modals/SwapVerifyTermsModal.js'
import { Airship, showError } from '../services/AirshipInstance'
import { type Theme, type ThemeProps, cacheStyles, withTheme } from '../services/ThemeContext.js'
import { Alert } from '../themed/Alert'
import { EdgeText } from '../themed/EdgeText.js'
import { ExchangeQuote } from '../themed/ExchangeQuoteComponent.js'
import { LineTextDivider } from '../themed/LineTextDivider'
import { SceneHeader } from '../themed/SceneHeader'

type OwnProps = {
  route: RouteProp<'exchangeQuote'>
}
type StateProps = {
  account: EdgeAccount,
  fromDenomination: string,
  fromWalletCurrencyName: string,
  pending: boolean,
  toDenomination: string,
  toWalletCurrencyName: string
}
type DispatchProps = {
  shift: (swapInfo: GuiSwapInfo, onApprove: () => void) => void,
  timeExpired: (swapInfo: GuiSwapInfo, onApprove: () => void) => void
}
type Props = StateProps & DispatchProps & ThemeProps & OwnProps

type State = {}

export class CryptoExchangeQuoteScreenComponent extends React.Component<Props, State> {
  calledApprove: true

  componentDidMount = () => {
    const { route, account, timeExpired } = this.props
    const { swapInfo, onApprove } = route.params
    const { pluginId } = swapInfo.quote
    const swapConfig = account.swapConfig[pluginId]

    logEvent('SwapQuote')
    swapVerifyTerms(swapConfig).then(result => {
      if (!result) timeExpired(swapInfo, onApprove)
    }, showError)
  }

  componentWillUnmount() {
    const { route } = this.props
    const { swapInfo } = route.params
    const { quote } = swapInfo

    if (!this.calledApprove) quote.close()
  }

  doShift = () => {
    const { shift, route } = this.props
    const { swapInfo, onApprove } = route.params
    this.calledApprove = true
    shift(swapInfo, onApprove)
  }

  renderTimer = () => {
    const { timeExpired, route } = this.props
    const { swapInfo, onApprove } = route.params
    const { expirationDate } = swapInfo.quote

    if (!expirationDate) return null
    return <CircleTimer timeExpired={() => timeExpired(swapInfo, onApprove)} expiration={expirationDate} />
  }

  showExplanationForEstimate = () => {
    Airship.show(bridge => (
      <ButtonsModal
        bridge={bridge}
        title={s.strings.estimated_exchange_rate}
        message={s.strings.estimated_exchange_rate_body}
        buttons={{ ok: { label: s.strings.string_ok } }}
      />
    ))
  }

  render() {
    const { account, fromDenomination, fromWalletCurrencyName, toDenomination, toWalletCurrencyName, pending, theme, route } = this.props
    const { swapInfo } = route.params
    const { fee, fromDisplayAmount, fromFiat, fromTotalFiat, toDisplayAmount, toFiat, quote, request } = swapInfo
    const { fiatCurrencyCode } = request.fromWallet
    const { pluginId } = quote
    const swapConfig = account.swapConfig[pluginId]
    const exchangeName = swapConfig.swapInfo.displayName
    const feePercent = div(quote.networkFee.nativeAmount, quote.fromNativeAmount, 2)
    const showFeeWarning = gte(feePercent, '0.05')
    const styles = getStyles(theme)
    return (
      <SceneWrapper background="theme">
        <SceneHeader withTopMargin title={s.strings.title_exchange} underline />
        <ScrollView>
          <LineTextDivider title={s.strings.fragment_send_from_label} lowerCased />
          {showFeeWarning && <Alert marginRem={[0, 1, 1.5, 1]} title={s.strings.transaction_details_fee_warning} type="warning" />}
          <ExchangeQuote
            cryptoAmount={fromDisplayAmount}
            currency={fromWalletCurrencyName}
            currencyCode={fromDenomination}
            fiatCurrencyAmount={fromFiat}
            fiatCurrencyCode={fiatCurrencyCode.replace('iso:', '')}
            isTop
            miningFee={fee}
            total={fromTotalFiat}
            walletId={request.fromWallet.id}
            walletName={request.fromWallet.name || ''}
            showFeeWarning={showFeeWarning}
          />
          <LineTextDivider title={s.strings.string_to_capitalize} lowerCased />
          <ExchangeQuote
            cryptoAmount={toDisplayAmount}
            currency={toWalletCurrencyName}
            currencyCode={toDenomination}
            fiatCurrencyAmount={toFiat}
            fiatCurrencyCode={request.toWallet.fiatCurrencyCode.replace('iso:', '')}
            walletId={request.toWallet.id}
            walletName={request.toWallet.name || ''}
          />
          <View style={styles.pluginRowPoweredByRow}>
            <EdgeText style={styles.footerText}>{s.strings.plugin_powered_by + ' '}</EdgeText>
            <FastImage style={styles.partnerIconImage} resizeMode="contain" source={{ uri: getSwapPluginIconUri(quote.pluginId, theme) }} />
            <EdgeText style={styles.footerText}>{' ' + exchangeName}</EdgeText>
          </View>
          {quote.isEstimate && (
            <Alert
              title={s.strings.estimated_quote}
              message={s.strings.estimated_exchange_message}
              type="warning"
              marginRem={[1.5, 1]}
              onPress={this.showExplanationForEstimate}
            />
          )}
          <Slider parentStyle={styles.slider} onSlidingComplete={this.doShift} disabled={pending} showSpinner={pending} />
          {this.renderTimer()}
          <View style={styles.spacer} />
        </ScrollView>
      </SceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  footerText: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  partnerIconImage: {
    aspectRatio: 1,
    width: theme.rem(0.75),
    height: theme.rem(0.75)
  },
  pluginRowPoweredByRow: {
    marginTop: theme.rem(1),
    flexDirection: 'row',
    justifyContent: 'flex-end',
    paddingRight: theme.rem(1),
    alignItems: 'center'
  },
  slider: {
    marginTop: theme.rem(2.5)
  },
  spacer: {
    height: theme.rem(8)
  }
}))

export const CryptoExchangeQuote = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    account: state.core.account,
    fromDenomination: state.cryptoExchange.fromWalletPrimaryInfo.displayDenomination.name,
    fromWalletCurrencyName:
      state.cryptoExchange.fromWalletId != null ? state.core.account.currencyWallets[state.cryptoExchange.fromWalletId].currencyInfo.displayName : '',
    pending: state.cryptoExchange.shiftPendingTransaction,
    toDenomination: state.cryptoExchange.toWalletPrimaryInfo.displayDenomination.name,
    toWalletCurrencyName:
      state.cryptoExchange.toWalletId != null ? state.core.account.currencyWallets[state.cryptoExchange.toWalletId].currencyInfo.displayName : ''
  }),
  dispatch => ({
    shift(swapInfo, onApprove) {
      dispatch(shiftCryptoCurrency(swapInfo, onApprove))
    },
    timeExpired(swapInfo, onApprove) {
      dispatch(exchangeTimerExpired(swapInfo, onApprove))
    }
  })
)(withTheme(CryptoExchangeQuoteScreenComponent))
