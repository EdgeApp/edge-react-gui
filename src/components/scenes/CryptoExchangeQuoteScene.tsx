import { div, gte } from 'biggystring'
import { EdgeAccount } from 'edge-core-js/types'
import * as React from 'react'
import { ScrollView, TouchableOpacity } from 'react-native'
import FastImage from 'react-native-fast-image'
import IonIcon from 'react-native-vector-icons/Ionicons'
import { sprintf } from 'sprintf-js'

import { exchangeTimerExpired, shiftCryptoCurrency } from '../../actions/CryptoExchangeActions'
import { lstrings } from '../../locales/strings'
import { connect } from '../../types/reactRedux'
import { EdgeSceneProps, NavigationBase } from '../../types/routerTypes'
import { GuiSwapInfo } from '../../types/types'
import { getSwapPluginIconUri } from '../../util/CdnUris'
import { getWalletName } from '../../util/CurrencyWalletHelpers'
import { logEvent } from '../../util/tracking'
import { NotificationSceneWrapper } from '../common/SceneWrapper'
import { ButtonsModal } from '../modals/ButtonsModal'
import { swapVerifyTerms } from '../modals/SwapVerifyTermsModal'
import { CircleTimer } from '../progress-indicators/CircleTimer'
import { Airship, showError } from '../services/AirshipInstance'
import { cacheStyles, Theme, ThemeProps, withTheme } from '../services/ThemeContext'
import { Alert } from '../themed/Alert'
import { EdgeText } from '../themed/EdgeText'
import { ExchangeQuote } from '../themed/ExchangeQuoteComponent'
import { LineTextDivider } from '../themed/LineTextDivider'
import { SceneHeader } from '../themed/SceneHeader'
import { Slider } from '../themed/Slider'

interface OwnProps extends EdgeSceneProps<'exchangeQuote'> {}

interface StateProps {
  account: EdgeAccount
  fromDenomination: string
  fromWalletCurrencyName: string
  pending: boolean
  toDenomination: string
  toWalletCurrencyName: string
}
interface DispatchProps {
  shift: (navigation: NavigationBase, swapInfo: GuiSwapInfo, onApprove: () => void) => Promise<void>
  timeExpired: (navigation: NavigationBase, swapInfo: GuiSwapInfo, onApprove: () => void) => Promise<void>
}
type Props = StateProps & DispatchProps & ThemeProps & OwnProps

interface State {}

export class CryptoExchangeQuoteScreenComponent extends React.Component<Props, State> {
  // @ts-expect-error
  calledApprove: true

  componentDidMount = () => {
    const { route, account, timeExpired, navigation } = this.props
    const { swapInfo, onApprove } = route.params
    const { pluginId } = swapInfo.quote
    const swapConfig = account.swapConfig[pluginId]

    logEvent('Exchange_Shift_Quote')
    swapVerifyTerms(swapConfig)
      .then(async result => {
        if (!result) await timeExpired(navigation, swapInfo, onApprove)
      })
      .catch(err => showError(err))
  }

  componentWillUnmount() {
    const { route } = this.props
    const { swapInfo } = route.params
    const { quote } = swapInfo

    if (!this.calledApprove) quote.close().catch(err => showError(err))
  }

  doShift = async () => {
    const { shift, route, navigation } = this.props
    const { swapInfo, onApprove } = route.params
    this.calledApprove = true
    await shift(navigation, swapInfo, onApprove)
  }

  renderTimer = () => {
    const { timeExpired, route, navigation } = this.props
    const { swapInfo, onApprove } = route.params
    const { expirationDate } = swapInfo.quote

    if (!expirationDate) return null
    return <CircleTimer timeExpired={async () => await timeExpired(navigation, swapInfo, onApprove)} expiration={expirationDate} />
  }

  handleForEstimateExplanation = async () => {
    await Airship.show<'ok' | undefined>(bridge => (
      <ButtonsModal
        bridge={bridge}
        title={lstrings.estimated_exchange_rate}
        message={lstrings.estimated_exchange_rate_body}
        buttons={{ ok: { label: lstrings.string_ok } }}
      />
    ))
  }

  handleCanBePartialExplanation = async () => {
    const { canBePartial, maxFulfillmentSeconds } = this.props.route.params.swapInfo.quote
    let canBePartialString: string | undefined
    if (canBePartial === true) {
      if (maxFulfillmentSeconds != null) {
        const t = Math.ceil(maxFulfillmentSeconds / 60)
        canBePartialString = sprintf(lstrings.can_be_partial_quote_with_max_body, t.toString())
      } else {
        canBePartialString = lstrings.can_be_partial_quote_body
      }
    }
    await Airship.show<'ok' | undefined>(bridge => (
      <ButtonsModal bridge={bridge} title={lstrings.can_be_partial_quote_title} message={canBePartialString} buttons={{ ok: { label: lstrings.string_ok } }} />
    ))
  }

  handlePoweredByTap = () => {
    this.props.navigation.navigate('exchangeSettings', {})
  }

  render() {
    const { account, fromDenomination, fromWalletCurrencyName, toDenomination, toWalletCurrencyName, pending, theme, route, navigation } = this.props
    const { swapInfo } = route.params
    const { fee, fromDisplayAmount, fromFiat, fromTotalFiat, toDisplayAmount, toFiat, quote, request } = swapInfo
    const { fiatCurrencyCode } = request.fromWallet
    const { pluginId } = quote
    const swapConfig = account.swapConfig[pluginId]
    const exchangeName = swapConfig.swapInfo.displayName
    const feePercent = div(quote.networkFee.nativeAmount, quote.fromNativeAmount, 2)
    const showFeeWarning = gte(feePercent, '0.05')
    const styles = getStyles(theme)

    return (
      <NotificationSceneWrapper hasTabs navigation={navigation} background="theme">
        {(gap, notificationHeight) => (
          <>
            <SceneHeader title={lstrings.title_exchange} underline withTopMargin />
            <ScrollView contentContainerStyle={[{ paddingBottom: notificationHeight }, styles.container]}>
              <LineTextDivider title={lstrings.fragment_send_from_label} lowerCased />
              {showFeeWarning && <Alert marginRem={[0, 1, 1.5, 1]} title={lstrings.transaction_details_fee_warning} type="warning" />}
              <ExchangeQuote
                cryptoAmount={fromDisplayAmount}
                currency={fromWalletCurrencyName}
                currencyCode={fromDenomination}
                fiatCurrencyAmount={fromFiat}
                fiatCurrencyCode={fiatCurrencyCode.replace('iso:', '')}
                isTop
                miningFee={fee}
                total={fromTotalFiat}
                walletId={request.fromWallet.id}
                walletName={getWalletName(request.fromWallet)}
                showFeeWarning={showFeeWarning}
              />
              <LineTextDivider title={lstrings.string_to_capitalize} lowerCased />
              <ExchangeQuote
                cryptoAmount={toDisplayAmount}
                currency={toWalletCurrencyName}
                currencyCode={toDenomination}
                fiatCurrencyAmount={toFiat}
                fiatCurrencyCode={request.toWallet.fiatCurrencyCode.replace('iso:', '')}
                walletId={request.toWallet.id}
                walletName={getWalletName(request.toWallet)}
              />
              <TouchableOpacity style={styles.pluginRowPoweredByRow} onPress={this.handlePoweredByTap}>
                <EdgeText style={styles.footerText}>{lstrings.plugin_powered_by_space + ' '}</EdgeText>
                <FastImage style={styles.partnerIconImage} resizeMode="contain" source={{ uri: getSwapPluginIconUri(quote.pluginId, theme) }} />
                <EdgeText style={styles.footerText}>{' ' + exchangeName}</EdgeText>
                <IonIcon name="chevron-forward" size={theme.rem(1)} color={theme.iconTappable} />
              </TouchableOpacity>
              {quote.isEstimate && (
                <Alert
                  title={lstrings.estimated_quote}
                  message={lstrings.estimated_exchange_message}
                  type="warning"
                  marginRem={[1, 1]}
                  onPress={this.handleForEstimateExplanation}
                />
              )}
              {quote.canBePartial === true && (
                <Alert
                  title={lstrings.can_be_partial_quote_title}
                  message={lstrings.can_be_partial_quote_message}
                  type="warning"
                  marginRem={[1, 1]}
                  onPress={this.handleCanBePartialExplanation}
                />
              )}

              <Slider parentStyle={styles.slider} onSlidingComplete={this.doShift} disabled={pending} showSpinner={pending} />
              {this.renderTimer()}
            </ScrollView>
          </>
        )}
      </NotificationSceneWrapper>
    )
  }
}

const getStyles = cacheStyles((theme: Theme) => ({
  container: {
    paddingTop: theme.rem(0.5)
  },
  footerText: {
    fontSize: theme.rem(0.75),
    color: theme.secondaryText
  },
  partnerIconImage: {
    aspectRatio: 1,
    width: theme.rem(0.75),
    height: theme.rem(0.75)
  },
  pluginRowPoweredByRow: {
    marginTop: theme.rem(1),
    flexDirection: 'row',
    justifyContent: 'flex-end',
    paddingRight: theme.rem(1),
    alignItems: 'center'
  },
  slider: {
    marginTop: theme.rem(0.5),
    marginBottom: theme.rem(1)
  },
  spacer: {
    height: theme.rem(8)
  }
}))

export const CryptoExchangeQuote = connect<StateProps, DispatchProps, OwnProps>(
  (state, { route: { params } }) => ({
    account: state.core.account,
    fromDenomination: state.cryptoExchange.fromWalletPrimaryInfo.displayDenomination.name,
    fromWalletCurrencyName:
      state.cryptoExchange.fromWalletId != null ? state.core.account.currencyWallets[state.cryptoExchange.fromWalletId].currencyInfo.displayName : '',
    pending: state.cryptoExchange.shiftPendingTransaction,
    toDenomination: state.cryptoExchange.toWalletPrimaryInfo.displayDenomination.name,
    toWalletCurrencyName:
      state.cryptoExchange.toWalletId != null ? state.core.account.currencyWallets[state.cryptoExchange.toWalletId].currencyInfo.displayName : ''
  }),
  dispatch => ({
    async shift(navigation, swapInfo, onApprove) {
      await dispatch(shiftCryptoCurrency(navigation, swapInfo, onApprove))
    },
    async timeExpired(navigation, swapInfo, onApprove) {
      await dispatch(exchangeTimerExpired(navigation, swapInfo, onApprove))
    }
  })
)(withTheme(CryptoExchangeQuoteScreenComponent))
