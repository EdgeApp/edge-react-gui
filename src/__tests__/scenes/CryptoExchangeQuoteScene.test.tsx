import { describe, expect, it } from '@jest/globals'
import { EdgeSwapInfo } from 'edge-core-js'
import * as React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { CryptoExchangeQuoteScreenComponent } from '../../components/scenes/CryptoExchangeQuoteScene'
import { getTheme } from '../../components/services/ThemeContext'
import { GuiSwapInfo } from '../../types/types'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

const dummySwapInfo: EdgeSwapInfo = {
  pluginId: '',
  displayName: '',
  supportEmail: ''
}

describe('CryptoExchangeQuoteScreenComponent', () => {
  it('should render with loading props', () => {
    const renderer = createRenderer()

    const fakeAccount: any = {
      swapConfig: {
        ChangeNow: {
          swapInfo: {
            pluginId: 'ChangeNow',
            displayName: 'ChangeNow',
            orderUri: 'ChangeNow.ChangeNow',
            supportEmail: 'ChangeNow@ChangeNow'
          }
        }
      }
    }

    const fakeRequest: any = {
      fromWallet: {
        fiatCurrencyCode: 'USD',
        name: ''
      },
      toWallet: {
        fiatCurrencyCode: 'USD',
        name: ''
      }
    }

    const swapInfo: GuiSwapInfo = {
      quote: {
        swapInfo: dummySwapInfo,
        request: fakeRequest,
        isEstimate: true,
        fromNativeAmount: '10000',
        toNativeAmount: '10000',
        networkFee: {
          currencyCode: 'BTC',
          nativeAmount: '1'
        },
        pluginId: 'ChangeNow',
        approve: async () => ({
          orderId: 'demo',
          transaction: {
            blockHeight: 500000,
            currencyCode: 'BTC',
            date: 1524486980,
            deviceDescription: 'iphone12',
            isSend: true,
            memos: [],
            metadata: {
              name: 'Crazy Person 2',
              category: 'Expense: Less Money',
              notes: 'Hell yeah! Here\'s a fish"',
              amountFiat: 36001.45
            },
            nativeAmount: '-321000000',
            networkFee: '2000',
            ourReceiveAddresses: ['receiveaddress3', 'receiveaddress4'],
            parentNetworkFee: '20001',
            signedTx: 'fiuwh34f98h3tiuheirgserg',
            txid: 'txid2',
            walletId: ''
          }
        }),
        close: async () => undefined
      },
      request: fakeRequest,
      fee: '1',
      fromDisplayAmount: '1',
      fromFiat: '1',
      fromTotalFiat: '1',
      toDisplayAmount: '1',
      toFiat: '1'
    }

    const actual = renderer.render(
      <CryptoExchangeQuoteScreenComponent
        {...fakeSceneProps('exchangeQuote', {
          swapInfo,
          onApprove: () => undefined
        })}
        account={fakeAccount}
        fromDenomination="BTC"
        fromWalletCurrencyName={{ fromDenomination: '' } as any}
        pending
        toDenomination="ETH"
        toWalletCurrencyName={{ fromDenomination: '' } as any}
        shift={async (swapInfo, onApprove) => {}}
        timeExpired={async (swapInfo, onApprove) => {}}
        theme={getTheme()}
      />
    )

    expect(actual).toMatchSnapshot()
  })
})
