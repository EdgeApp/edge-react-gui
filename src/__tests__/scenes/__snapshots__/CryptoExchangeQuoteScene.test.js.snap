// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoExchangeQuoteScreenComponent should render with loading props 1`] = `
<SceneWrapper
  background="theme"
>
  <Memo(SceneHeaderComponent)
    title="Exchange"
    underline={true}
    withTopMargin={true}
  />
  <ScrollView>
    <Memo(LineTextDividerComponent)
      lowerCased={true}
      title="From"
    />
    <WithTheme(ExchangeQuoteComponent)
      currency={
        Object {
          "fromDenomination": "",
        }
      }
      currencyCode="BTC"
      fiatCurrencyCode="USD"
      isTop={true}
      showFeeWarning={false}
      walletName=""
    />
    <Memo(LineTextDividerComponent)
      lowerCased={true}
      title="To"
    />
    <WithTheme(ExchangeQuoteComponent)
      currency={
        Object {
          "fromDenomination": "",
        }
      }
      currencyCode="ETH"
      fiatCurrencyCode="USD"
      walletName=""
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "marginTop": 22,
          "paddingRight": 22,
        }
      }
    >
      <WithTheme(EdgeTextComponent)
        style={
          Object {
            "color": "#A4C7DF",
            "fontSize": 17,
          }
        }
      >
        Powered by 
      </WithTheme(EdgeTextComponent)>
      <FastImage
        resizeMode="contain"
        source={
          Object {
            "uri": "https://content.edge.app/exchangeIcons/ChangeNow/icon.png",
          }
        }
        style={
          Object {
            "aspectRatio": 1,
            "height": 17,
            "width": 17,
          }
        }
      />
      <WithTheme(EdgeTextComponent)
        style={
          Object {
            "color": "#A4C7DF",
            "fontSize": 17,
          }
        }
      >
         ChangeNow
      </WithTheme(EdgeTextComponent)>
    </View>
    <Alert
      marginRem={
        Array [
          1.5,
          1,
        ]
      }
      message="The amount above is an estimate. This exchange may result in less funds received than quoted."
      onPress={[Function]}
      title="Estimated Quote"
      type="warning"
    />
    <WithTheme(SliderComponent)
      disabled={true}
      onSlidingComplete={[Function]}
      parentStyle={
        Object {
          "marginTop": 56,
        }
      }
      showSpinner={true}
    />
    <View
      style={
        Object {
          "height": 180,
        }
      }
    />
  </ScrollView>
</SceneWrapper>
`;
